### ISSUE TITLE
Specialization for faster count(:Array{Bool})

### ISSUE THREAD

**Issue by jakobnissen** [JuliaLang/julia#34059]
So I noticed recently that Python's `numpy` is quite a bit faster at counting boolean values than Julia is. On my laptop, `numpy.count_nonzero` takes 1.1 µs for 10,000 values whereas `sum(x)` in Julia takes 2.1 µs. So, we're almost 2x slower here.

However, we can write a specialized function for summing/counting an `Array{Bool}`:
```
function count(x::Array{Bool})
    n = 0
    chunks = length(x) ÷ sizeof(UInt)
    mask = 0x0101010101010101 % UInt
    GC.@preserve x begin
        ptr = Ptr{UInt}(pointer(x))
        for i in 1:chunks
            n += count_ones(unsafe_load(ptr, i) & mask)
        end
    end
    for i in 8chunks+1:length(x)
        n += x[i]
    end
    n
end
```
This function takes 300 ns, making us comfortably faster than Numpy.
Now, the question is: Is it worth it to put this specialized function in Base, or is this unnecessary bloat? I'd say that counting boolean arrays are a relatively common task and worth optimizing. If so, I'll pitch a PR.

Related issue: #34046

**Comment by haampie** [comment]
For what it's worth, the issue boils down to:

```julia
using BenchmarkTools
get_mask(::Type{UInt64}) = 0x0101010101010101
get_mask(::Type{UInt32}) = 0x01010101
get_mask(::Type{UInt8})  = 0x01

function mwe(x::Vector{T}) where T
    cnt = 0
    mask = get_mask(T)
    for y in x
        cnt += count_ones(y & mask)
    end
    cnt
end
```

```julia
julia> @btime mwe(x) setup=(x = rand(UInt64, 2^11)) # 16384 byte
  416.417 ns (0 allocations: 0 bytes)
8286

julia> @btime mwe(x) setup=(x = rand(UInt32, 2^12)) # 16384 byte
  2.224 μs (0 allocations: 0 bytes)
8177

julia> @btime mwe(x) setup=(x = rand(UInt8, 2^14)) # 16384 byte
  11.415 μs (0 allocations: 0 bytes)
8258
```

It would be nice if the compiler generated the same code for UInt8, UInt32 and UInt64 here...

**Comment by jakobnissen** [comment]
Fixed by #34060
